### ISSUE TITLE
RFC: EntityNetworkQueueHandler.

### ISSUE THREAD

**Issue by filfat** [JSPrismarine/JSPrismarine#121]
# RFC
## EntityNetworkQueueHandler
*TL;DR: The idea is to introduce a queue system for handling outbound, inbound packets and their respective handlers. It should also be completely multi-thread safe.*

### Introduction
The current network stack associates packet writing, reading and handling directly to entities and players, just like other projects have discovered this is really slow and inefficient. Both from a performance and development perspective. We desperately need to find a better solution, that's why I'm proposing a per-entity queue system.

While RakNet does handle a queue of incoming and outgoing packets this does not account for decoding/encoding and handling of those packets which is where we see the actual thread-blocking.

#### Entity & Player
Each entity will have a property called networkQueue with the type `EntityNetworkQueueHandler`, except for the player class which will instead have a `PlayerNetworkQueueHandler` that's extending `EntityNetworkQueueHandler`. At this initial stage the `PlayerNetworkQueueHandler` won't have any specific functionality.

#### EntityNetworkQueueHandler
The goal of this class is to handle the actual queue for the entity. For that we'll need several functions:
##### `private interface NetworkQueueEntry { timestamp: number, packet: DataPacket, priority: number, timeout: number }`
* `timestamp` Time stamp of when the entry got added in ticks.
* `packet` the packet to be processed by the worker pool.
* `priority` Number between 0-10.
* `timeout` Number of ticks before discarding the entry.
##### `public void reset[Inbound | Outbound]Queue(void)`
This one is self explanatory, it will completely reset the queue. It's really only useful for an entity world change or server reload.
##### `public DataPacket pop[Inbound | Outbound]Queue(void)`
This will return the first relevant (see `calculatePacketRelavancy` for more information regarding how relevancy is calculated) packet from the queue and remove it from the internal queue system.
##### `public void push[Inbound | Outbound]Queue(packet: DataPacket, priority: number, timeout?: number)`
* `priority` number between 0-10, where lower is higher priority.
* `timeout` number of ticks before discarding the packet from the queue.

This will push a packet to the internal queue and run the private function `calculatePacketRelavancy`.
##### `private void calculate[Inbound | Outbound]PacketRelavancy(void)`
This will update the internal queue based on packet priority and timeout.
If a timeout is met the packet is removed from the internal queue with a `silly` log output.

### Example usage (pseudo code)
```typescript
function sendPlayerMovementToPlayer(target: Player) {
	const pk = PlayerMovementPacket();
	pk.player = this.runtimeId;
	pk.x = this.x;
	pk.y = this.y;
	pk.z = this.z;
	target.getPlayerNetworkQueueHandler().pushOutboundQueue(
		pk,
		target.getPriorityFromPlayerDistance(this),
		5 // Timeout
	);
}
```

### Final words
This system will later be expanded upon by another RFC regarding a pool of per-entity (or player) workers in which each entity processes each packet decoding/encoding & handling in parallel.

**Comment by Bavfalcon9** [comment]
In response to RakNet,
I believe it makes more sense to implement this directly into RakNet.

From your proposal, 
RakNet should have it's own `NetworkQueue` that directly interfaces with RakNet connections (so we don't have to do this in JSP). This class would be extensible and `EntityNetworkQueueHandler` would extend it. This class would send and receive packets as you suggested in your proposal. However in order to do this, we would need to get rid of (or change) how Connections are handled in RakNet, because Implementing this without changing (or seperating) the `Connection` class in raknet would be redundant.

Another opinion, I think from the API side of things, we should implement this behind a simple `sendPacket()` (or related) this way when we change the internal api of how queues will function, we're not causing API break's towards plugins. So a plugin would be able to:

```ts
const player: Player = server.getPlayer('SomeUserName');
player.sendPacket(myPacket, { EntryOptions });
```
And in the player class we would have the sendPacket, do our queuing. (hopefully this makes sense).

**Comment by HerryYT** [comment]
> In response to RakNet,
> I believe it makes more sense to implement this directly into RakNet.
> 
> From your proposal,
> RakNet should have it's own `NetworkQueue` that directly interfaces with RakNet connections (so we don't have to do this in JSP). This class would be extensible and `EntityNetworkQueueHandler` would extend it. This class would send and receive packets as you suggested in your proposal. However in order to do this, we would need to get rid of (or change) how Connections are handled in RakNet, because Implementing this without changing (or seperating) the `Connection` class in raknet would be redundant.
> 
> Another opinion, I think from the API side of things, we should implement this behind a simple `sendPacket()` (or related) this way when we change the internal api of how queues will function, we're not causing API break's towards plugins. So a plugin would be able to:
> 
> ```ts
> const player: Player = server.getPlayer('SomeUserName');
> player.sendPacket(myPacket, { EntryOptions });
> ```
> 
> And in the player class we would have the sendPacket, do our queuing. (hopefully this makes sense).

Plugins shouldn't access network things directly, we should provide an api on the player class and then on the network handler we should add all the hidden network functions that are executed on the player class functions i think

**Comment by Bavfalcon9** [comment]
That's what I mean with the `sendPacket`

**Comment by filfat** [comment]
@Bavfalcon9  I don’t think that plugins should directly invoke sendPacket unless it’s a modded client with custom packets (in fact we should actively block known already internally handled packets from being sent through sendPacket by a plugin).

We should instead provide a versioned API that supports everything a plugin wants to do. That way plugins won’t break upon internal changes, neither will they miss handling things correctly.
In other words instead of `sendPacket(movePacket)` they’ll do something like `player.moveTo(x, y, z)`.

**Comment by HerryYT** [comment]
**This prototype is a little bit broken**
How could an **entity** have a NetworkQueue if:
- Entities are spawned by sending a packet to the client and they're all managed with the entity unique Id that you put in the packet to spawn them. Entities also don't have an inbound connection (client -> server).
- When constructing NetworkQueueHandler, i should base the player on it, let me explain... when someone connects, RakNet handles all his login sequence and finally constructs the NetworkQueueHandler class, but then, how am i supposed to extend this class to a PlayerNetworkQueueHandler class? i should directly construct PlayerNetworkQueue class that extends the NetworkQueueHandler class (and then pass it as an argument to the player class), it would mix a little bit RakNet with Minecraft related code and this thing isn't that good, but we may instead make an interface so people can implement their own NetworkQueueHandler with not only Minecraft related functions.

**Comment by HerryYT** [comment]
However, i am proposing some changes in order to fix all those problems:
- NetworkQueueHandler is extended by PlayerConnection that is found in network root folder, all the folders: network/packet, handler, type will be moved to network/minecraft/packet, handler, type to clean up network folder (also we must move away packet binary stream, basically i just want PlayerConnection on network folder root).
- For entities we will make an EntitiyManager class with all the functions in order to manage the entity via his unique id, and sync the moving / actions of them for every player outbound queue
